### DISCUSSION TITLE
Memory Usage

### DISCUSSION THREAD

**Discussion by cfengno1** [block-hczhai/block2-preview#145] [Q&A] [answered]
Hi Huanchen,

I have a few quesitons about the memory usage in block2.

1. I notice that there is a parameter ``min_mpo_mem" for dynamically loading MPO. Can I do the same for MPS? 

2. According to the document, the parameter ``stack_mem" stands for the memory used for storing renormalized operators (in bytes). Does it mean that all environment tensors are stored in memory instead of on disk?

3. I notice three additional parameters that save memory: fused_contraction_multiplication, fused_contraction_rotation, lowmem_numerical_transform. Can you help explain what is fused contraction and numerical transform used here?

Thank you so much!

Best,
Feng

**Comment by hczhai** [comment]
1. MPS is by default dynamically loaded into memory.
2. Only renormalized operators that are relevant to the current iteration is loaded into memory.
3. As shown in the Fig. 6 of the paper *J. Chem. Phys.* **154**, 224116 (2021), "contraction" is Fig. 6(a), "rotation" is Fig. 6 (b), and "multiplication" is Fig. 6(c). "Numerical transform" is explained in https://block2.readthedocs.io/en/latest/theory/spatial.html#middle-site-transformation. When you are not doing DMRG for quantum chemistry Hamiltonian using the conventional implementation, there is no "numerical transform" step.

**Reply by cfengno1** [reply]
Hi Huanchen,

THank you for the quick reply!

Regarding the contraction, rotation and multiplication, my queston is what does "fused" mean? How does it save memory?

Best,
Feng

**Reply by hczhai** [reply, marked as answer]
Fused X and Y means ``opY(opX(A, B), C)`` is implemented as ``opXY(A, B, C)`` without storing the intermediate ``opX(A, B)`` in memory. Since these options may conflict with other more effective optimizations, it is very likely for a specific usage case there is no saving of the peak memory during the whole calculation. Normally the default option is recommended.

**Reply by cfengno1** [reply]
Hi Huanchen,

I have a follow-up question 

The document says that stack_mem_ratio is "The fraction of stack space occupied by the main stacks" and stack_mem is "The memory used for storing renormalized operators". Does it mean that among the stack_mem memory, stack_mem_ratio of it is used for main stacks which store things like MPO and MPS, and the rest of it stores renormalized operators?

Thank you so much!

Best,
Feng

**Reply by hczhai** [reply]
MPO tensors will be in kept in dynamically allocated memory or disk depending on whether ``min_mpo_mem == True``, so it has nothing to do with ``stack_mem``. MPS tensors will be in disk most of the time, so it is also not related to ``stack_mem``. These MPO and MPS tensors are not the main consumer of the DMRG memory cost. You can find the technical details of the memory model used in ``block2`` in https://block2.readthedocs.io/en/latest/api/global.html, but it is not going to be helpful for you to solve any practical memory issues.

In principle, changing the value of ``stack_mem_ratio`` or ``stack_mem`` will not save you any memory cost during DMRG. For practical guide of setting a suitable ``stack_mem`` parameter, you can have a look at https://github.com/block-hczhai/block2-preview/issues/10#issuecomment-1104680616.

**As I mentioned in a previous [reply](https://github.com/block-hczhai/block2-preview/discussions/151), if you indeed have memory related issue when running ``block2``, please provide a concrete setting, including the description of the system, the input and output file, the available resources (CPU models, number of nodes, and memory available in each node) in your computational environment, what memory related error message you get, and whether the same input works at smaller number of sites or bond dimensions. We can then do a case study, and I will provide suggestions based on the details of your case. The case does not have to be your real research project. It can be any similar setting that can reproduce your memory problem. This way we can avoid discussing technical details or abstract concepts that are irrelevant for your usage case.**

**Reply by cfengno1** [reply]
Hi Huanchen,

Thank you for your reply!

Let me be more clear of the issue. I have a large MPS bond dimension calculation that cost 200-300G memory but I am not sure of the exact memory cost. The output print is 
```
Dmem = 9.43 GB (12%) | Imem = 138 KB (90%) | Hmem = 107 GB | Wmem = 591 MB | Pmem = 0 B
```

Now I want to use mpi to run this job on a few nodes which have 128G memory and 40 cores each. I have tried to set in the slurm
```
#SBATCH --nodes=2
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=40
#SBATCH --mem=120G
```
and in the python script
```
driver = DMRGDriver(scratch="./tmp", symm_type=SymmetryTypes.SAny | SymmetryTypes.CPX , n_threads=40, stack_mem=120 << 30, mpi=True)
```
, but it still meets the memory shortage issue.

I have the following questions:

1. Is the above setup correct for using MPI? E.g. n_threads=40 is correct?
2. If I use more MPI processor, should the calculation cost less memory on each node and thus solve the memory issue?
3. Is it a good practice to set stack_mem equal to the total memory allocated by slurm?
4. Can one read out the peak memory usage from the print out?


Thank you again!

Best,
Feng

**Reply by hczhai** [reply]
1 & 2. Simply setting ``mpi=True`` will not work unless it is a standard quantum chemistry Hamiltonian. To see how to do MPI parallelization for model Hamiltonians, you need to provide the complete input script.
3. ``stack_mem`` should be set to the memory available per processor (if you have only one MPI processor per node, this is memory per node).
4. To print the peak memory memory usage (on processor 0) at each sweep iteration, you can add the following before calling ``driver.dmrg`` and the ``iprint`` argument of ``driver.dmrg`` should be ``iprint=2``:

```python3
def f(name, iprint):
    import resource
    if (driver.mpi is None or driver.mpi.rank == 0) and name == "DMRG::sweep::iter.start":
        print("%10s" % driver.bw.b.Parsing.to_size_string(
            1024 * resource.getrusage(resource.RUSAGE_SELF).ru_maxrss), end='', flush=True)

driver.set_callback(f)
```

**Reply by cfengno1** [reply]
The python script is
```
from pyblock2.driver.core import DMRGDriver, SymmetryTypes, MPOAlgorithmTypes
import numpy as np

LX, LY = 40, 6
L = LX * LY
J1 = 1
t1 = 3
J2 = 0.06
t2 = t1 * (J2 ** 0.5)
theta = np.pi/12
Nh = 30
NE = L - Nh
TWO_S = 0
NTHREADS = 40
Nmem = 120

driver = DMRGDriver(scratch="./tmp", symm_type=SymmetryTypes.SAny | SymmetryTypes.CPX , n_threads=NTHREADS, stack_mem=Nmem << 30, restart_dir="./MPS", clean_scratch=False, mpi=True)

driver.set_symmetry_groups("U1Fermi", "U1") # electron number, doublon number, total spin, total spin
Q = driver.bw.SX

# [Part A] Set states and matrix representation of operators in local Hilbert space
site_basis, site_ops = [], []

for k in range(L):
    basis = [(Q(0, 0), 1), (Q(1, 1), 1), (Q(1, -1), 1)] # [0 up down]
    ops = {
        "": np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]]),        # identity
        "c": np.array([[0, 0, 0], [1, 0, 0], [0, 0, 0]]),       # alapha+
        "d": np.array([[0, 1, 0], [0, 0, 0], [0, 0, 0]]),  # alpha
        "C": np.array([[0, 0, 0], [0, 0, 0], [1, 0, 0]]),       # beta+
        "D": np.array([[0, 0, 1], [0, 0, 0], [0, 0, 0]]),       # beta
        "z": np.array([[0, 0, 0], [0, 0.5, 0], [0, 0, -0.5]]),       # Sz
        "p": np.array([[0, 0, 0], [0, 0, 1], [0, 0, 0]]),       # S+
        "m": np.array([[0, 0, 0], [0, 0, 0], [0, 1, 0]]),       # S-
        "n": np.array([[0, 0, 0], [0, 1, 0], [0, 0, 1]]),  # electron number
    }
    site_basis.append(basis)
    site_ops.append(ops)

# [Part B] Set Hamiltonian terms
driver.initialize_system(n_sites=L, vacuum=Q(0, 0), target=Q(NE, TWO_S), hamil_init=False)
driver.ghamil = driver.get_custom_hamiltonian(site_basis, site_ops)

b = driver.expr_builder()
la = np.loadtxt("Lattice.dat",delimiter=",",dtype=int)

for i in range(0,la.shape[0]):
    bd = la[i][0]
    s1 = la[i][1] - 1
    s2 = la[i][2] - 1

    if bd == 1:
        th = theta
        t = t1
        J = J1
    else:
        th = 0
        t = t2
        J = J2
    b.add_term("pm", [s1, s2], J/2 * np.exp(-2j*th))
    b.add_term("pm", [s2, s1], J/2 * np.exp(2j*th))
    b.add_term("zz", [s1, s2], J)
    b.add_term("nn", [s1, s2], -J/4)
    b.add_term("cd", [s1, s2], -t*np.exp(-1j*th))
    b.add_term("cd", [s2, s1], -t*np.exp(1j*th))
    b.add_term("CD", [s1, s2], -t*np.exp(1j*th))
    b.add_term("CD", [s2, s1], -t*np.exp(-1j*th))

mpo = driver.get_mpo(b.finalize(adjust_order=True, fermionic_ops="cdCD"), algo_type=MPOAlgorithmTypes.FastBipartite, iprint=1)

mps = driver.load_mps(tag="KET", nroots=1)

sweep_start = 0
sweep_num = 12
forward = True

bondD = [10000] * 6 + [12000] * 6
energy = driver.dmrg(mpo, mps, n_sweeps=sweep_num, bond_dims=bondD,
        noises=[0], thrds=[1e-10] * sweep_num, dav_max_iter=30, iprint=4, forward=forward, sweep_start=sweep_start)
print("DMRG energy = %20.15f (per site = %10.6f)" % (energy, energy / L))
```
where I loaded a previously calculated MPS.

The slurm script is 
```
#!/bin/bash
### SHEBANG IS REQUIRED
    
#SBATCH --time=1-00:00:00
#SBATCH --partition=asterix
#SBATCH --nodes=6
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=10 
#SBATCH --mem=120G
        
        
### USER COMMANDS TO EXECUTE
module purge 
module load anaconda3
conda activate block2
module load openmpi

python Tri_cont.py >> output.txt
```

Lattice.dat is attached.
[Lattice.dat.zip](https://github.com/user-attachments/files/18816607/Lattice.dat.zip)

Thanks!
Feng

**Reply by hczhai** [reply]
Thanks for providing the details.

To make your script work with MPI parallelization, you can do the following modification to the "Part B":

```python3
# [Part B] Set Hamiltonian terms
driver.initialize_system(n_sites=L, vacuum=Q(0, 0), target=Q(NE, TWO_S), hamil_init=False)
driver.ghamil = driver.get_custom_hamiltonian(site_basis, site_ops)

b = driver.expr_builder()
la = np.loadtxt("Lattice.dat",delimiter=",",dtype=int)

for i in range(0,la.shape[0]):
    bd = la[i][0]
    s1 = la[i][1] - 1
    s2 = la[i][2] - 1

    if min(s1, s2) % driver.mpi.size != driver.mpi.rank:
        continue

    if bd == 1:
        th = theta
        t = t1
        J = J1
    else:
        th = 0
        t = t2
        J = J2
    b.add_term("pm", [s1, s2], J/2 * np.exp(-2j*th))
    b.add_term("pm", [s2, s1], J/2 * np.exp(2j*th))
    b.add_term("zz", [s1, s2], J)
    b.add_term("nn", [s1, s2], -J/4)
    b.add_term("cd", [s1, s2], -t*np.exp(-1j*th))
    b.add_term("cd", [s2, s1], -t*np.exp(1j*th))
    b.add_term("CD", [s1, s2], -t*np.exp(1j*th))
    b.add_term("CD", [s2, s1], -t*np.exp(-1j*th))

mpo = driver.get_mpo(b.finalize(adjust_order=True, fermionic_ops="cdCD"), algo_type=MPOAlgorithmTypes.FastBipartite, iprint=1)

driver.mpi.barrier()
mps = driver.load_mps(tag="KET", nroots=1)
driver.mpi.barrier()
```

In your SLURM script, the last line should be

```bash
export MKL_NUM_THREADS=1
export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK
export PYTHONUNBUFFERED=1
mpirun --map-by ppr:$SLURM_TASKS_PER_NODE:node:pe=$SLURM_CPUS_PER_TASK python3 -u Tri_cont.py >> output.txt
```

The ``NTHREADS`` variable in your script should match ``#SBATCH --cpus-per-task=10`` in the SLURM script (ideally, you should set this to the total number of cores in the node, which is 40). The ``Nmem`` variable in your script should match ``#SBATCH --mem=120G`` in the SLURM script.

The added line ``if min(s1, s2) % driver.mpi.size != driver.mpi.rank: continue`` is needed to make sure each MPI processor get a unique set of Hamiltonian terms. The amount of memory (per node) that you can save largely depends on how to divide the terms into different nodes. (For example, ``if s1 % driver.mpi.size != driver.mpi.rank: continue`` will likely be worse for your lattice.) The best way to write this line is system and lattice dependent.

After you add the above line, you can check the printed ``MPO bond dimension`` to see the potential saving in memory (the memory cost is linear with respect to the MPO bond dimension per node). For example, for your case, when you use one node, the ``MPO bond dimension = 98``. When you use two nodes, you will see ``MPO bond dimension = 58`` and ``MPO bond dimension = 50`` for node 0 and node 1, so you may expect a up to 41% reduction in memory cost per node. When you use four nodes, you will see ``MPO bond dimension = 34`` for node 0/1/2/3 so you may expect a up to 65% reduction in memory cost per node.

**Reply by cfengno1** [reply]
Hi Huanchen,

Just a follow-up question. 

In the SLURM script you recommended:
export MKL_NUM_THREADS=1
export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK

Now if I want to set **n_mkl_threads = 4** in the DMRGDriver, should I set:
export MKL_NUM_THREADS=4
export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK / 4 ?

Thank you so much!

Best,
Feng

**Reply by hczhai** [reply]
You can do
```bash
unset MKL_NUM_THREADS
export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK
```

**Comment by cfengno1** [comment]
Dear Huanchen,

Can you explain to me in the modified code, why are there two driver.mpi.barrier:
```
driver.mpi.barrier()
mps = driver.load_mps(tag="KET", nroots=1)
driver.mpi.barrier()
```

However, in your code in https://github.com/hczhai/2d-hubbard-dmrg-2024/blob/master/05-8x16-N120/05-8x16-N120-x2-re1.py,
only one is there.

Thank you!

Best,

**Reply by hczhai** [reply]
You can delete the first ``driver.mpi.barrier()`` and it will not affect the result.
